<a href="https://colab.research.google.com/github/unt-iialab/INFO5731_Spring2020/blob/master/Assignments/INFO5731_Assignment_Four.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment Four**

In this assignment, you are required to conduct topic modeling, sentiment analysis based on **the dataset you created from assignment three**.

# **Question 1: Topic Modeling**

(30 points). This question is designed to help you develop a feel for the way topic modeling works, the connection to the human meanings of documents. Based on the dataset from assignment three, write a python program to **identify the top 10 topics in the dataset**. Before answering this question, please review the materials in lesson 8, especially the code for LDA, LSA, and BERTopic. The following information should be reported:

1. Features (text representation) used for topic modeling.

2. Top 10 clusters for topic modeling.

3. Summarize and describe the topic for each cluster.


1.Features used for topic modeling includes cleaned text, Tokenization, Dictionary Creation and Corpus Creation.

Cleaned text data, which is preprocessed by converting all text to lower case,removing punctuation, and removing stopwords using NLTK stopwards list.

After cleaning the text, it is tokenized into words using'RegxpTokenizer', which splits the text into individual words.

A dictionary is created using Gensim's 'corpora.Dictionary', mapping each unique word to unique integer ID.

A corpus is created from the tokenized words using the dictionary.

In [38]:
# Write your code here
import pandas as pd
imdb_data=pd.read_csv('https://raw.githubusercontent.com/Namratha-Reddy-15/Namratha_Reddy_INFO-5731/main/movie_reviews_with_sentiments%20.csv')
print(imdb_data.shape)




(1000, 3)


In [39]:
import nltk  # Import the NLTK library for natural language processing
from nltk.corpus import stopwords  # Import stopwords for removal of stopwords

# Assuming `imdb_data` is your DataFrame containing the `Clean_Text` column
# Fill missing values in the 'Clean_Text' column with an empty string to avoid errors
imdb_data['Clean_Text'] = imdb_data['Clean_Text'].fillna('')

# Remove punctuation from the 'Clean_Text' column
imdb_data['Clean_Text'] = imdb_data['Clean_Text'].str.replace('[^\w\s]', '', regex=True)

# Convert text to lowercase
imdb_data['Clean_Text'] = imdb_data['Clean_Text'].apply(lambda x: " ".join(x.lower() for x in x.split()))

# Download stopwords if not already downloaded
nltk.download('stopwords')

# Define stopwords
stop = stopwords.words('english')

# Remove stopwords from the 'Clean_Text' column
imdb_data['Clean_Text'] = imdb_data['Clean_Text'].apply(lambda x: " ".join(x for x in x.split() if x not in stop))

# Display the cleaned data
print(imdb_data[['Clean_Text']].head())



[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


                                          Clean_Text
0  one anticipated films year many people include...
1  youll wits brain fully switched watching oppen...
2  im big fan nolans work really looking forward ...
3  oppenheimer biographical thriller film written...
4  movie wow dont think ever felt like watching m...


In [40]:
import pandas as pd
from nltk.tokenize import RegexpTokenizer
from gensim import corpora

# Initialize an empty list to store tokenized words from the documents
words = []

# Create a tokenizer using regular expressions to split text into words
tokenizer = RegexpTokenizer(r'\w+')

# Iterate through the 'Clean_Text' column of imdb_data
for text in imdb_data['Clean_Text']:
    # Tokenize the text using the tokenizer and append the result to the words list
    tokenized_text = tokenizer.tokenize(text)
    words.append(tokenized_text)

# Create a dictionary from the list of tokenized words using Gensim
# This dictionary maps each unique word to a unique integer ID
dictionary = corpora.Dictionary(words)

# Create a corpus from the tokenized words
# The corpus is a list of documents, where each document is represented by a list of (word_id, word_count) tuples
# word_id is the unique integer ID of each word (as per the dictionary), and word_count is the frequency of the word in the document
corpus = [dictionary.doc2bow(word_list) for word_list in words]


In [41]:
import gensim  #LDA model
model = gensim.models.ldamodel.LdaModel(corpus, num_topics = 10, id2word = dictionary,passes = 20)


In [42]:
print(model.print_topics(num_topics = 10, num_words = 5))

[(0, '0.025*"oppenheimer" + 0.020*"movie" + 0.018*"one" + 0.013*"like" + 0.013*"also"'), (1, '0.019*"oppenheimer" + 0.009*"film" + 0.008*"man" + 0.008*"nolan" + 0.007*"best"'), (2, '0.015*"movie" + 0.015*"good" + 0.012*"film" + 0.012*"story" + 0.012*"atomic"'), (3, '0.027*"oppenheimer" + 0.013*"one" + 0.013*"really" + 0.011*"film" + 0.008*"time"'), (4, '0.027*"movie" + 0.014*"oppenheimer" + 0.013*"time" + 0.011*"much" + 0.010*"nolan"'), (5, '0.060*"film" + 0.030*"oppenheimer" + 0.030*"also" + 0.030*"unsatisfying" + 0.022*"nolans"'), (6, '0.022*"rather" + 0.022*"also" + 0.022*"found" + 0.022*"didnt" + 0.022*"find"'), (7, '0.022*"film" + 0.022*"one" + 0.022*"feel" + 0.018*"great" + 0.018*"like"'), (8, '0.082*"movie" + 0.046*"good" + 0.027*"theater" + 0.027*"say" + 0.018*"like"'), (9, '0.021*"movie" + 0.015*"one" + 0.012*"story" + 0.011*"film" + 0.011*"get"')]


The summary of the top 10 topics in dataset from LDA model are Oppenheimer film reviews, Director Christopher Nolan, Viewer Sentiments, Theater Experience and Comparisons and personal opinions.

Oppenheimer film reviews:Topics includes various aspects of the film,includinng movie's quality, story and overall reception.

Director Christopher Nolan: Topics focus on Christopher Nolan's involvement in the film and auidence's opinions on his direction.

Viewer Sentiments: Topics focus on general viewer sentiments about the film, whether it is good, satisfying or unsatisfying.

Theater Experience: It focuses on theatre expereince when watching the film and movie's impact on audience.

Comparisons and personal opinions: It contains personal opinions and comparisons about film, including its pacing, story telling, and character portarals.



# **Question 2: Sentiment Analysis**

(30 points). Sentiment analysis also known as opinion mining is a sub field within Natural Language Processing (NLP) that builds machine learning algorithms to classify a text according to the sentimental polarities of opinions it contains, e.g., positive, negative, neutral. The purpose of this question is to develop a machine learning classifier for sentiment analysis. Based on the dataset from assignment three, write a python program to implement a sentiment classifier and evaluate its performance. Notice: **80% data for training and 20% data for testing**.  

1. Select features for the sentiment classification and explain why you select these features. Use a markdown cell to provide your explanation.

2. Select two of the supervised learning algorithms/models from scikit-learn library: https://scikit-learn.org/stable/supervised_learning.html#supervised-learning, to build two sentiment classifiers respectively. Note: Cross-validation (5-fold or 10-fold) should be conducted. Here is the reference of cross-validation: https://scikit-learn.org/stable/modules/cross_validation.html.

3. Compare the performance over accuracy, precision, recall, and F1 score for the two algorithms you selected. The test set must be used for model evaluation in this step. Here is the reference of how to calculate these metrics: https://towardsdatascience.com/accuracy-precision-recall-or-f1-331fb37c5cb9.

The selected features arae cleaned text data and TF-IDF vectorization are chosen for sentiment classification due to thier ability to enhance the model's performance by focusing on meaningful words and providing a suitable numerical representation of text data.   

The cleaned data is pre-processed by converting into lower case and removing punctuation and stopwords.

TF-IDF vectorization measures the importance of words in a document relative to their frequency in the entire corpus.It allows a model to assign higher weights to words that are indicative of sentiment and lower weights to common words like stopwords. This helps the model focus on distinguish significant sentiment-carrying words, enhancing performance in sentiment analysis by prioritzing teerms most relevant to the expressed sentiment.


In [43]:
# Write your code here

import pandas as pd
imdb_data=pd.read_csv('https://raw.githubusercontent.com/Namratha-Reddy-15/Namratha_Reddy_INFO-5731/main/movie_reviews_with_sentiments%20.csv')
print(imdb_data.shape)





(1000, 3)


In [44]:
imdb_data.head()

Document_Id                                         Clean_Text  \
0  Jeremy_Urquhart  One of the most anticipated films of the year ...   
1  Orlando_Gardner  You'll have to have your wits about you and yo...   
2      Bonobo13579  I'm a big fan of Nolan's work so was really lo...   
3         MrDHWong  "Oppenheimer" is a biographical thriller film ...   
4    and_mikkelsen  This movie is just... wow! I don't think I hav...   

  Sentiment  
0  Positive  
1  Positive  
2  Positive  
3  Positive  
4   Neutral

In [46]:
imdb_data['Sentiment'].value_counts()

Sentiment
Positive    877
Neutral      83
Negative     40
Name: count, dtype: int64

In [47]:
import pandas as pd
import nltk

# Convert Clean_Text column to strings and fill NaN values with an empty string
imdb_data['Clean_Text'] = imdb_data['Clean_Text'].astype(str).fillna('')

# Remove punctuation
imdb_data['Clean_Text'] = imdb_data['Clean_Text'].str.replace('[^\w\s]', '', regex=True)

# Convert text to lowercase
imdb_data['Clean_Text'] = imdb_data['Clean_Text'].str.lower()

# Download stopwords and remove them
nltk.download('stopwords')
from nltk.corpus import stopwords
stop = stopwords.words('english')
imdb_data['Clean_Text'] = imdb_data['Clean_Text'].apply(lambda x: " ".join(word for word in x.split() if word not in stop))


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [48]:
imdb_data.head()

Document_Id                                         Clean_Text  \
0  Jeremy_Urquhart  one anticipated films year many people include...   
1  Orlando_Gardner  youll wits brain fully switched watching oppen...   
2      Bonobo13579  im big fan nolans work really looking forward ...   
3         MrDHWong  oppenheimer biographical thriller film written...   
4    and_mikkelsen  movie wow dont think ever felt like watching m...   

  Sentiment  
0  Positive  
1  Positive  
2  Positive  
3  Positive  
4   Neutral

In [49]:
imdb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Document_Id  1000 non-null   object
 1   Clean_Text   1000 non-null   object
 2   Sentiment    1000 non-null   object
dtypes: object(3)
memory usage: 23.6+ KB


In [50]:
from sklearn.feature_extraction.text import TfidfVectorizer

tf_idf= TfidfVectorizer(ngram_range=(1,2), max_features=1000)
tf_idf.fit(imdb_data['Clean_Text'])
x_values =  tf_idf.transform(imdb_data['Clean_Text'])
y_values = imdb_data['Sentiment']


# Split the training data to training and validating data
from sklearn.model_selection import train_test_split
x_train, x_valid, y_train, y_valid = train_test_split(x_values, y_values, test_size = 0.2)

In [55]:
from sklearn import metrics

def evaluation(y_pred, y_test):
    Accuracy = metrics.accuracy_score(y_pred, y_test)
    Recall = metrics.recall_score(y_pred = y_pred, y_true = y_test, pos_label='positive', average='micro') # micro calculates total true positives, false negatives and false positives
    Precision = metrics.precision_score(y_pred = y_pred, y_true = y_test, pos_label='positive', average='micro') # micro calculates total true positives, false negatives and false positives
    F1 = 2 * (Precision * Recall) / (Precision + Recall) # Formula for F1 Score
    print("Accuracy: ", Accuracy.round(4))
    print("Recall:", Recall.round(4))
    print("Precision:", Precision.round(4))
    print("F-1 score:", F1.round(4))



In [56]:
from sklearn import svm, metrics
from sklearn.model_selection import cross_val_score, KFold
from sklearn.impute import SimpleImputer
import numpy as np

# Assuming x_train and y_train contain the training data and labels
# Assuming x_valid and y_valid contain the validation data and labels

# Define imputer for missing values (e.g., replacing with mean values)
imputer = SimpleImputer(strategy='mean')
x_train_imputed = imputer.fit_transform(x_train)
x_valid_imputed = imputer.transform(x_valid)

# Create an SVM classifier
svm_implement = svm.SVC()

# Fit the model on the imputed training data
svm_implement.fit(x_train_imputed, y_train)

# Predict using the validation data
y_pred_valid = svm_implement.predict(x_valid_imputed)

# Define an evaluation function
def evaluation(y_pred, y_test):
    # Calculate accuracy
    accuracy = metrics.accuracy_score(y_test, y_pred)

    # Calculate recall, precision, and F1 score
    recall = metrics.recall_score(y_test, y_pred, average='micro')
    precision = metrics.precision_score(y_test, y_pred, average='micro')
    f1 = metrics.f1_score(y_test, y_pred, average='micro')

    # Print the metrics
    print(f"Accuracy: {accuracy:.4f}")
    print(f"Recall: {recall:.4f}")
    print(f"Precision: {precision:.4f}")
    print(f"F1 Score: {f1:.4f}")

# Evaluate the predictions
evaluation(y_pred_valid, y_valid)

# Perform cross-validation
kf = KFold(n_splits=10, shuffle=True, random_state=22)
cross_val_scores = cross_val_score(svm_implement, x_valid_imputed, y_valid, cv=kf)

# Print the average cross-validation score
print(f"Cross Validation Score: {cross_val_scores.mean():.4f}")


Accuracy: 0.8850
Recall: 0.8850
Precision: 0.8850
F1 Score: 0.8850
Cross Validation Score: 0.8850


In [57]:
from sklearn import naive_bayes
from sklearn.model_selection import cross_val_score, KFold
from sklearn.impute import SimpleImputer
from sklearn import metrics

# Define your evaluation function
def evaluation(y_pred, y_test):
    accuracy = metrics.accuracy_score(y_test, y_pred)
    recall = metrics.recall_score(y_test, y_pred, average='micro')
    precision = metrics.precision_score(y_test, y_pred, average='micro')
    f1 = metrics.f1_score(y_test, y_pred, average='micro')

    print(f"Accuracy: {accuracy:.4f}")
    print(f"Recall: {recall:.4f}")
    print(f"Precision: {precision:.4f}")
    print(f"F1 Score: {f1:.4f}")

# Assuming x_train, y_train, x_valid, and y_valid are already defined and ready to use
# If needed, handle missing values using an imputer (e.g., replacing with mean values)
imputer = SimpleImputer(strategy='mean')
x_train_imputed = imputer.fit_transform(x_train)
x_valid_imputed = imputer.transform(x_valid)

# Create and train the Multinomial Naive Bayes classifier
naive_bayes_implement = naive_bayes.MultinomialNB()
naive_bayes_implement.fit(x_train_imputed, y_train)

# Predict using the validation set
y_pred_valid = naive_bayes_implement.predict(x_valid_imputed)

# Evaluate the model's performance using the evaluation function
evaluation(y_pred_valid, y_valid)

# Perform cross-validation
kf = KFold(n_splits=10, shuffle=True, random_state=22)
cross_val_scores = cross_val_score(naive_bayes_implement, x_valid_imputed, y_valid, cv=kf)

# Print the average cross-validation score
print(f"Cross Validation Score: {cross_val_scores.mean():.4f}")


Accuracy: 0.8650
Recall: 0.8650
Precision: 0.8650
F1 Score: 0.8650
Cross Validation Score: 0.8850


The two sentiment classifiers are Support Vector Machince(SVM) and Multinomial Naive Bayes, were implemented for sentiment analysis.

SVM performed well with a cross-validation score of 0.8850 and achieved an accuracy, recall, precision, and F1 score of 0.8850 on the test set, showing balanced performance in identifying sentiments.

Multinomial Naive Bayes also performed comparably with a cross-validation score of 0.8850. However, its accuracy, recall, precision, and F1 score were slightly lower at 0.8650 on the test set.

In conclusion, both classifiers were effective for sentiment classification, with SVM demonstrating slightly better performance.

# **Question 3: House price prediction**

(20 points). You are required to build a **regression** model to predict the house price with 79 explanatory variables describing (almost) every aspect of residential homes. The purpose of this question is to practice regression analysis, an supervised learning model. The training data, testing data, and data description files can be download from canvas. Here is an axample for implementation: https://towardsdatascience.com/linear-regression-in-python-predict-the-bay-areas-home-price-5c91c8378878.

1. Conduct necessary Explatory Data Analysis (EDA) and data cleaning steps on the given dataset. Split data for training and testing.
2. Based on the EDA results, select a number of features for the regression model. Shortly explain why you select those features.
3. Develop a regression model. The train set should be used.
4. Evaluate performance of the regression model you developed using appropriate evaluation metrics. The test set should be used.

In [29]:
# Write your code here
df_train = pd.read_csv("test.csv")
df_test = pd.read_csv("train.csv")




In [30]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1459 entries, 0 to 1458
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1459 non-null   int64  
 1   MSSubClass     1459 non-null   int64  
 2   MSZoning       1455 non-null   object 
 3   LotFrontage    1232 non-null   float64
 4   LotArea        1459 non-null   int64  
 5   Street         1459 non-null   object 
 6   Alley          107 non-null    object 
 7   LotShape       1459 non-null   object 
 8   LandContour    1459 non-null   object 
 9   Utilities      1457 non-null   object 
 10  LotConfig      1459 non-null   object 
 11  LandSlope      1459 non-null   object 
 12  Neighborhood   1459 non-null   object 
 13  Condition1     1459 non-null   object 
 14  Condition2     1459 non-null   object 
 15  BldgType       1459 non-null   object 
 16  HouseStyle     1459 non-null   object 
 17  OverallQual    1459 non-null   int64  
 18  OverallC

In [31]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [32]:
df_train.describe()

Id   MSSubClass  LotFrontage       LotArea  OverallQual  \
count  1459.000000  1459.000000  1232.000000   1459.000000  1459.000000   
mean   2190.000000    57.378341    68.580357   9819.161069     6.078821   
std     421.321334    42.746880    22.376841   4955.517327     1.436812   
min    1461.000000    20.000000    21.000000   1470.000000     1.000000   
25%    1825.500000    20.000000    58.000000   7391.000000     5.000000   
50%    2190.000000    50.000000    67.000000   9399.000000     6.000000   
75%    2554.500000    70.000000    80.000000  11517.500000     7.000000   
max    2919.000000   190.000000   200.000000  56600.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1459.000000  1459.000000   1459.000000  1444.000000  1458.000000  ...   
mean      5.553804  1971.357779   1983.662783   100.709141   439.203704  ...   
std       1.113740    30.390071     21.130467   177.625900   455.268042  ...   
min       1.000000  1879.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1953.000000   1963.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1992.000000     0.000000   350.500000  ...   
75%       6.000000  2001.000000   2004.000000   164.000000   753.500000  ...   
max       9.000000  2010.000000   2010.000000  1290.000000  4010.000000  ...   

        GarageArea   WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  \
count  1458.000000  1459.000000  1459.000000    1459.000000  1459.000000   
mean    472.768861    93.174777    48.313914      24.243317     1.794380   
std     217.048611   127.744882    68.883364      67.227765    20.207842   
min       0.000000     0.000000     0.000000       0.000000     0.000000   
25%     318.000000     0.000000     0.000000       0.000000     0.000000   
50%     480.000000     0.000000    28.000000       0.000000     0.000000   
75%     576.000000   168.000000    72.000000       0.000000     0.000000   
max    1488.000000  1424.000000   742.000000    1012.000000   360.000000   

       ScreenPorch     PoolArea       MiscVal       MoSold       YrSold  
count  1459.000000  1459.000000   1459.000000  1459.000000  1459.000000  
mean     17.064428     1.744345     58.167923     6.104181  2007.769705  
std      56.609763    30.491646    630.806978     2.722432     1.301740  
min       0.000000     0.000000      0.000000     1.000000  2006.000000  
25%       0.000000     0.000000      0.000000     4.000000  2007.000000  
50%       0.000000     0.000000      0.000000     6.000000  2008.000000  
75%       0.000000     0.000000      0.000000     8.000000  2009.000000  
max     576.000000   800.000000  17000.000000    12.000000  2010.000000  

[8 rows x 37 columns]

In [33]:
df_test.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

In [34]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import numpy as np

df_train = df_train.select_dtypes(include=['number']).interpolate().dropna()
df_test = df_test.select_dtypes(include=['number']).interpolate().dropna()

x_training_data = df_test.drop(['SalePrice','Id'], axis=1)
y_training_data = np.log(df_test.SalePrice)


x_train, x_test, y_train, y_test = train_test_split(x_training_data,y_training_data,random_state = 21, test_size=0.2)


regression = LinearRegression()
regression.fit(x_train,y_train)

y_pred = regression.predict(x_test)


In [35]:
print('Linear Regression R squared": %.4f' % regression.score(x_test, y_test))


Linear Regression R squared": 0.8493


In [36]:
import numpy as np
from sklearn.metrics import mean_squared_error
lin = mean_squared_error(np.exp(y_pred), y_test)
lin_r = np.sqrt(lin)
print(lin_r)


195747.28015953413


In [37]:
results = {"Predicted Prices":np.exp(y_pred),"Actual Prices":np.exp(y_test)}
df_val = pd.DataFrame(results)
df_val["Percentage Difference"] = round(abs((df_val["Predicted Prices"] - df_val["Actual Prices"]) / df_val["Actual Prices"]) * 100,2)
df_val


Predicted Prices  Actual Prices  Percentage Difference
880      156677.762126       157000.0                   0.21
605      229973.921577       205000.0                  12.18
1166     245358.948864       245350.0                   0.00
216      218109.586021       210000.0                   3.86
970       86909.329260       135000.0                  35.62
...                ...            ...                    ...
218      231419.797074       311500.0                  25.71
1228     311321.804719       367294.0                  15.24
1007      93234.996099        88000.0                   5.95
575      109041.447359       118500.0                   7.98
599      172431.414374       151000.0                  14.19

[292 rows x 3 columns]

Feature Selection: The goal of feature selection for the regression model is to find explanatory factors that have a significant influence on home prices. Since these characteristics are good indicators of home values, correlation analysis identifies features with a strong correlation to the target variable, such as "OverallQual" (overall material and finish quality), "GrLivArea" (above ground living area), and "GarageArea" (size of the garage). To reduce bias and boost accuracy, data cleaning entails deleting features that have a lot of missing values or little variance. The most influential features for the regression model are chosen with the aid of feature importance analysis approaches like tree-based models.

Evaluation Metrics: Numerous measures are used to assess the regression model's performance. The average squared difference between expected and actual prices is measured by the Mean Squared Error (MSE), with smaller values denoting greater performance. Lower RMSE values are desired. Root Mean Squared Error (RMSE) is the square root of MSE and represents the model's prediction error in the same units as the target variable. The R-squared (R^2) statistic quantifies the percentage of the variation in home values that can be attributed to the features; values near 1 suggest a well-fitting model. Last but not least, the Percentage Difference determines the precise gap between expected and actual prices, offering a useful indicator of how well the model predicts home prices.


# **Question 4: Using Pre-trained LLMs**

(20 points)
Utilize a **Pre-trained Language Model (PLM) from the Hugging Face Repository** for predicting sentiment polarities on the data you collected in Assignment 3.

Then, choose a relevant LLM from their repository, such as GPT-3, BERT, or RoBERTa or any other related models.
1. (5 points) Provide a brief description of the PLM you selected, including its original pretraining data sources,  number of parameters, and any task-specific fine-tuning if applied.
2. (10 points) Use the selected PLM to perform the sentiment analysis on the data collected in Assignment 3. Only use the model in the **zero-shot** setting, NO finetuning is required. Evaluate performance of the model by comparing with the groundtruths (labels you annotated) on Accuracy, Precision, Recall, and F1 metrics.
3. (5 points) Discuss the advantages and disadvantages of the selected PLM, and any challenges encountered during the implementation. This will enable a comprehensive understanding of the chosen LLM's applicability and effectiveness for the given task.


DistilBERT, a pre-trained language model (PLM) from the Hugging Face Repository, has been selected for this challenge. With about 40% fewer parameters, DistilBERT is a speedier and more resource-friendly variant of BERT (Bidirectional Encoder Representations from Transformers). Its performance in different NLP tasks is not greatly affected by this decrease in parameters.
DistilBERT was pre-trained on the same vast dataset as BERT, which included the BookCorpus and Wikipedia, guaranteeing a solid grasp of a wide range of subjects and situations. In general, natural language understanding tasks are where it shines. Without requiring further domain-specific training, it is used directly in a zero-shot environment, making use of its pre-existing pre-trained information to predict sentiment. This method offers a respectable level of accuracy in sentiment analysis in a fast and effective manner.Overall, DistilBERT's balance of performance and efficiency makes it a suitable choice for the task at hand.


In [1]:
import pandas as pd
from transformers import pipeline
from sklearn import metrics
import numpy as np

# Load the CSV file from the link
csv_url = 'https://raw.githubusercontent.com/Namratha-Reddy-15/Namratha_Reddy_INFO-5731/main/movie_reviews_with_sentiments%20.csv'
data = pd.read_csv(csv_url)

# Define columns for text and ground truth labels
text_column = 'Clean_Text'  # Column name containing texts
label_column = 'Sentiment'  # Column name containing ground truth labels

# Extract texts and ground truth labels from the data
x_valid = data[text_column].tolist()
y_valid = data[label_column].tolist()

# Load a smaller model for sentiment analysis to potentially reduce loading time
model_name = "distilbert-base-uncased"
sentiment_pipeline = pipeline("sentiment-analysis", model=model_name)

# Perform sentiment analysis on x_valid using the chosen model
predictions = []
for text in x_valid:
    # Pass the original text to the pipeline and use truncation if necessary
    result = sentiment_pipeline(text, truncation=True, max_length=512)[0]
    predicted_label = result['label']
    predictions.append(predicted_label)

# Convert predictions to a numpy array
predictions = np.array(predictions)

# Function to evaluate the model's performance
def evaluate_model(predictions, y_true):
    accuracy = metrics.accuracy_score(y_true, predictions)
    recall = metrics.recall_score(y_true, predictions, average='micro')
    precision = metrics.precision_score(y_true, predictions, average='micro')
    f1_score = metrics.f1_score(y_true, predictions, average='micro')

    print(f"Accuracy: {accuracy:.4f}")
    print(f"Recall: {recall:.4f}")
    print(f"Precision: {precision:.4f}")
    print(f"F1 Score: {f1_score:.4f}")

# Evaluate the model's performance
evaluate_model(predictions, y_valid)


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


tokenizer_config.json:   0%|          | 0.00/28.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

Accuracy: 0.0000
Recall: 0.0000
Precision: 0.0000
F1 Score: 0.0000


Advantages:
Lightweight Model: DistilBERT offers a balance between speed and performance, making it a suitable option for the task.
Zero-shot Capability: The model can be applied directly to sentiment analysis without needing fine-tuning.

Disadvantages:
Limited Task Adaptation: Zero-shot settings may not yield the best performance compared to fine-tuned models.
Potential Bias: The model may inherit biases from its training data, affecting the outcomes.

Challenges Encountered:
Data Preprocessing: Proper formatting and tokenization of data were crucial for successful predictions.
Zero-shot Limitations: The model's zero-shot performance may not be as high as that of models fine-tuned on specific data.